### DISCUSSION TITLE
[Question][Hw2]

### DISCUSSION THREAD

**Discussion by seeifsalem** [prosyslab-classroom/cs348-information-security#684] [Q&A]
Name: Mohamed Seifeddine Salem

Hello,
I am having a hard time debugging the CBC encryption because this is what I stumble upon
![image](https://user-images.githubusercontent.com/73144051/229342886-d31fb827-70eb-4364-8f5f-d90a7d4ae1bd.png)
is there any way to obtain the actual characters instead of these ���� ?

**Comment by KAIST-JongchanPark** [comment]
Why don't you make a your own printer function and use it during encryption?

**Comment by seeifsalem** [comment]
> Why don't you make a your own printer function and use it during encryption?

But how would I know what is the expected result ?

**Comment by KAIST-JongchanPark** [comment]
You can use an external tool like `xxd`

**Comment by seeifsalem** [comment]
I don't understand how it would make sense for me to have the same two last characters ("wu") but having a mistake in the middle. We are using CBC so the encryption depends on the previously encoded block. If any of my encoded blocks was wrong, then the last encoded blocks can't be correct.

**Comment by KAIST-JongchanPark** [comment]
You may catch the concept of enc/dec, but your implementation might have a bug. Please debug your code carefully.

**Comment by leejy12** [comment]
@seeifsalem Have you checked if the expected file and your output have the same size?

**Comment by seeifsalem** [comment]
> @seeifsalem Have you checked if the expected file and your output have the same size?

how would I check that ? I mean it seems to me that they have the same number of characters

**Comment by leejy12** [comment]
Redirect your output from encrypting to another file like this:

```
./aes enc test/key.txt test/iv.txt test/helloworld.txt > test/helloworld.output
```
and compare the file sizes of `helloworld.expected` and `helloworld.output`.

Or, you can just use `wc`:

```
./aes enc test/key.txt test/iv.txt test/helloworld.txt | wc -c
```

This will count the number of characters. The expected value is 13 for helloworld.

**Comment by seeifsalem** [comment]
> Redirect your output from encrypting to another file like this:
> 
> ```
> ./aes enc test/key.txt test/iv.txt test/helloworld.txt > test/helloworld.output
> ```
> 
> and compare the file sizes of `helloworld.expected` and `helloworld.output`.
> 
> Or, you can just use `wc`:
> 
> ```
> ./aes enc test/key.txt test/iv.txt test/helloworld.txt | wc -c
> ```
> 
> This will count the number of characters. The expected value is 13 for helloworld.

I just checked, I have the same number of characters but I think I solved my problem.
